### PR TITLE
Google Cloud SDK for running Espresso tests

### PR DESCRIPTION
**Pull request by JaimeToca**
Hello Thyrlian, thanks for all the work you have put into this repo :+1: . This was such a good source of learning for me when I started with Docker/Android. I've used this Android SDK image for a while and found myself wanting to run UI test on Google Cloud device farm (Firebase Test Lab). Since I've extended this image and added Google Cloud SDK for myself I though someone else could benefit from it too.

Let me know what you think and if it makes sense for you ! If so, I could add a section in the README for how to run espresso tests step by step.

Cheers,
Jaime

### CODE REVIEW

**Comment by thyrlian** [comment]
Hi @JaimeToca, thanks a lot for your contribution 😃

It makes tons of sense to add support for other tools, especially for those popular ones, such as Firebase Test Lab.  However, Firebase Test Lab doesn't come for free, thus there are many people who choose other solutions.  This base image should only contain tools that are mandatory for minimal Android projects.  But as said, this tool is still a popular choice for some users, I'd suggest adding this one as a dedicated sub-image, like this one for [vnc](https://github.com/thyrlian/AndroidSDK/tree/master/android-sdk/vnc), then everyone could pick whatever they prefer.  Furthermore, it would be great if you could elaborate on how to use it in README.

Thanks again!  Good job!

**Comment by JaimeToca** [comment]
Hi @thyrlian !! You're right and it makes sense :+1: , so I'll add a new dedicated sub-image like the vnc one and step-by-step description to the README.  I do have experience with other device farms, perhaps if people are interested I could add some tools/script in the future? But for now I'll focus on Google Cloud SDK :+1: 

Thanks for your response !

**Comment by thyrlian** [comment]
That's a great idea, thanks man!

**Comment by JaimeToca** [comment]
Changes done. Let me know what you think @thyrlian !

Cheers !

**Comment by thyrlian** [comment]
Gracias @JaimeToca! Thanks for your effort. Sorry that I'm busy with some family obligations these days, will probably only have time to review and test your changes next weekend, hope you could understand.

Just one thought from my side at the moment, what do you think if changing all the names where it has been mentioned as **Google Cloud SDK** to **Firebase Test Lab**? IMHO, Google Cloud SDK is a bunch of stuff, to do this and that via Google Cloud services. And what we really offer, is the integration of Firebase Test Lab. With the name of Firebase Test Lab, the intention is clear, for testing purpose, and would be also easier for others to search this image by keywords.

**Comment by JaimeToca** [comment]
Hello @thyrlian no worries, take your time :+1:  ! Sure, I'll change the names I think Firebase Test Lab will clarify everything :+1: 

Thanks a lot,
Cheers

File: README.md [hunk 1]
```diff
@@ -573,6 +573,49 @@ Don't worry about `adbkey` or `adbkey.pub` under `/.android`, not required.
...
+
+* Now let's pull and run the Docker image with Google Cloud SDK
+
+```bash
+# pull the image
+docker pull thyrlian/android-firebase-test-lab
+# run the container passing the service account Json (at least the first time) as external volume
+docker run -d -p 2222:22 -v $(pwd)/sdk:/opt/android-sdk -v /path/on/host/auth.json:/auth.json thyrlian/android-firebase-test-lab
+```
+* Log in and run your tests
```
Comment (thyrlian) [inline README.md:590]: You can actually combine both code snippets into one section :)

File: README.md [hunk 2]
```diff
@@ -573,6 +573,49 @@ Don't worry about `adbkey` or `adbkey.pub` under `/.android`, not required.
...
+
+```bash
+# check the running process and get a interactive shell
+Docker ps
+Docker exec -it <container-id> /bin/bash
+
+# check Google Cloud is installed properly
+gcloud info --format="value(installation.sdk_root)"
```
Comment (thyrlian) [inline README.md:598]: This would be useful for testing and debugging.  I suggest moving this to a dedicated `version_inspector.sh` file inside the `firebase_test_lab` directory, and build this image, run this script with TravisCI (just add to `.travis.yml` configuration).

File: README.md [hunk 3]
```diff
@@ -573,6 +573,49 @@ Don't worry about `adbkey` or `adbkey.pub` under `/.android`, not required.
...
+# authenticate and create any config that you want
+gcloud auth activate-service-account -q --key-file myservicefile.json
+
+# list available devices to run your tests
+gcloud firebase test android models list
```
Comment (thyrlian) [inline README.md:603]: It makes sense to show some list data example here, especially the table schema part.

For instance:

```
┌───────────────────┬────────────────────┬──────────────────────────────────────┬──────────┬─────────────┬─────────────────────────┬───────────────┐
│      MODEL_ID     │        MAKE        │              MODEL_NAME              │   FORM   │  RESOLUTION │      OS_VERSION_IDS     │      TAGS     │
├───────────────────┼────────────────────┼──────────────────────────────────────┼──────────┼─────────────┼─────────────────────────┼───────────────┤
│ Nexus9            │ HTC                │ Nexus 9                              │ VIRTUAL  │ 2048 x 1536 │ 21,22,23,24,25          │               │
│ NexusLowRes       │ Generic            │ Low-resolution MDPI phone            │ VIRTUAL  │  640 x 360  │ 23,24,25,26,27,28,29,30 │ beta=30       │
│ OnePlus3T         │ OnePlus            │ OnePlus 3T                           │ PHYSICAL │ 1920 x 1080 │ 26                      │               │
└───────────────────┴────────────────────┴──────────────────────────────────────┴──────────┴─────────────┴─────────────────────────┴───────────────┘
```

And later, for the `$DEVICES` definition, you could refer here.

File: README.md [hunk 4]
```diff
@@ -573,6 +573,49 @@ Don't worry about `adbkey` or `adbkey.pub` under `/.android`, not required.
...
+
+# run your tests (make sure you have previosly assemble your app and test apk's)
+APK="--app=debug.apk --test=androidTest.apk"
+TYPE="instrumentation"
+DEVICES="--device model=Pixel2,version=27,locale=en,orientation=portrait"
```
Comment (thyrlian) [inline README.md:608]: Just a question: I guess you can run on multiple devices together, that's why you name it as plural, is this correct? :)

File: README.md [hunk 5]
```diff
@@ -573,6 +573,49 @@ Don't worry about `adbkey` or `adbkey.pub` under `/.android`, not required.
...
+
+# list available devices to run your tests
+gcloud firebase test android models list
+
+# run your tests (make sure you have previosly assemble your app and test apk's)
+APK="--app=debug.apk --test=androidTest.apk"
+TYPE="instrumentation"
+DEVICES="--device model=Pixel2,version=27,locale=en,orientation=portrait" 
+RESULT_DIR="build-result"
+gcloud firebase test android run $APK $DEVICES --type=$TYPE --results-dir=$RESULT_DIR/
```
Comment (thyrlian) [inline README.md:610]: You can remove the appending `/` from `$RESULT_DIR/`.

Comment (thyrlian) [reply README.md:610]: When I run this command, I got an error:

```
ERROR: (gcloud.firebase.test.android.run) No project specified. Please add --project PROJECT_ID to the command line or first run
  $ gcloud config set project PROJECT_ID
```

It seems like you also need to guide users to create a project and config it in [Firebase Console](https://console.firebase.google.com/).

Comment (thyrlian) [reply README.md:610]: BTW, I can't really test this one, Firebase tells me that I can only choose the **Blaze Plan (Pay as you go)** plan but not the **Spark Plan (Free)**. :(

> Since your Google Cloud Platform project has billing enabled, you'll be on the Firebase Blaze plan (pay as you go).

File: README.md [hunk 6]
```diff
@@ -573,6 +573,49 @@ Don't worry about `adbkey` or `adbkey.pub` under `/.android`, not required.
...
+# check Google Cloud is installed properly
+gcloud info --format="value(installation.sdk_root)"
+# authenticate and create any config that you want
+gcloud auth activate-service-account -q --key-file myservicefile.json
+
+# list available devices to run your tests
+gcloud firebase test android models list
+
+# run your tests (make sure you have previosly assemble your app and test apk's)
+APK="--app=debug.apk --test=androidTest.apk"
```
Comment (thyrlian) [inline README.md:606]: Maybe you can tell users how to build apk and test apk as well:

```shell
# to build the application apk
./gradlew :<your_module>:assemble
# apk will be generated at: <your_module>/build/outputs/apk/<build_type>

# to build the test apk
./gradlew :<your_module>:assembleAndroidTest
# apk will be generated at: <your_module>/build/outputs/apk/androidTest/<build_type>
```

File: image_publisher.sh [hunk 7]
```diff
@@ -13,8 +13,10 @@ DOCKER_HUB_ACCOUNT=thyrlian
 DOCKER_FILE_NAME=Dockerfile
 MAIN_IMAGE_NAME=android-sdk
 MAIN_IMAGE_DIR=android-sdk
-SUB_IMAGE_NAME=android-sdk-vnc
-SUB_IMAGE_DIR=vnc
+SUB_IMAGE_VNC_NAME=android-sdk-vnc
+SUB_IMAGE_VNC_DIR=vnc
+SUB_IMAGE_FIREBASE_TEST_LAB_NAME=android-firebase-test-lab
```
Comment (thyrlian) [inline image_publisher.sh:18]: It can follow the naming convention: `android-sdk-firebase-test-lab`

File: image_publisher.sh [hunk 8]
```diff
@@ -13,8 +13,10 @@ DOCKER_HUB_ACCOUNT=thyrlian
...
+SUB_IMAGE_FIREBASE_TEST_LAB_DIR=firebase_test_lab
```
Comment (thyrlian) [inline image_publisher.sh:19]: Following the naming convention, the directory name should use hyphen instead of underscore: `firebase-test-lab`

**Review by thyrlian — CHANGES_REQUESTED** [review]
Thanks again, all look good, but require some minor changes (except for one place: `gcloud firebase test android run...` has encountered an error, needs further configuration on Firebase side, please see review comments for more details).

Looking forward to your changes! :)

**Comment by JaimeToca** [comment]
Hey @thyrlian thanks a lot for your time 👌. I'll find some time this weekend and make the changes suggested.

Thanks !

**Comment by thyrlian** [comment]
Thank you Jaime for your contribution and time!

Meanwhile, I'm trying the Firebase Blaze plan, to see if it works on my side.  What an expensive review 😂

**Comment by JaimeToca** [comment]
Oh no 😓, I could upload a video and share with you.

**Comment by thyrlian** [comment]
Don't worry, it should be just a few $, not too much, just want to double-check that it works on multiple machines/accounts/conditions, so we're safe :)
Already on it, testing now

**Comment by thyrlian** [comment]
Sorry @JaimeToca, after passing `--project PROJECT_ID`, I got another authentication issue, could you please help me a hand?

```
ERROR: (gcloud.firebase.test.android.run) Unable to access the test environment catalog: ResponseError 403: Not authorized for project XXX
```

I found one answer on [stackoverflow](https://stackoverflow.com/questions/46917461/gcloud-firebase-android-test-unable-to-access-the-test-environment-catalog-res).  But I have already given the editor role to the service account, according to your guideline, don't know what to do now.

Thanks.

**Comment by thyrlian** [comment]
Wait, my bad (or Firebase's)...

Just found after creating / linking project in Firebase, it automatically created a new firebase user in my project in Google Cloud Platform. I'll try to grant permission for this user as well, and generate a new json key.

**Comment by thyrlian** [comment]
Hmm, I've created a new key for the new firebase user, and used it in the container, still got the same error...

**Comment by JaimeToca** [comment]
🤔🤔🤔 thats strange. I'll check it out tonight and get back to you, perhaps I forgot a step in the process

File: README.md (hunk 4 above)
Comment (JaimeToca) [reply README.md:608]: yes :) , I'll add info about splitting an Espresso test-suite into multiple shards and execute them in parallel.

File: README.md (hunk 5 above)
Comment (JaimeToca) [reply README.md:610]: oh :( my project is set up with "pay as you go" plan, I thought It could be done with the free plan tbh. 
Regarding --project PROJECT_ID I would probably did it time ago and forgot about it. I'll update the info srry

File: README.md (hunk 4 above)
Comment (thyrlian) [reply README.md:608]: Cool, thanks!

File: README.md (hunk 5 above)
Comment (thyrlian) [reply README.md:610]: The command says: you need this project option for the first run.

**Comment by thyrlian** [comment]
Regarding access permission:

1st: after account creation, you can't configure role in **Service Accounts** anymore, but you can find it in **IAM** -> **PERMISSIONS**.

2nd: I've added even **Owner** permission to the newly created Firebase account, but I still got the same error :(  I don't think there is any delay of role configuration taking effect.

**Comment by thyrlian** [comment]
Update: it works for me now.
I just need to re-authenticate, with the new key created: `gcloud auth activate-service-account -q --key-file /auth.json`.

**Comment by thyrlian** [comment]
Another update:

* I removed both **Editor** and **Owner** roles from the newly created Firebase account, re-authenticated, re-ran the test -> not immediately, but got not authorized error in about a minute.

* Added **Editor** role back for the Firebase account, re-ran test, everything is fine.  So no need for **Owner** role.

Finally, it's working :)

**Comment by JaimeToca** [comment]
Awesome @thyrlian, glad is working for you ! Huge thanks for your review and time.

I'll update the PR soon !

**Comment by thyrlian** [comment]
No problem Jaime, take your time :)

**Comment by thyrlian** [comment]
I've just given another successful try, please find below the minimum configuration in GCP and Firebase:

* Create a project in [Google Cloud Platform](https://console.cloud.google.com/cloud-resource-manager) if you haven't created one yet.
* Create a project in [Firebase](https://console.firebase.google.com/):
  * Choose the recently created Google Cloud Platform project to add Firebase services to it.
  * Confirm Firebase billing plan.
* Go to **IAM & Admin** -> **Service Accounts** in [Google Cloud Platform](https://console.cloud.google.com/iam-admin/serviceaccounts):
  * Edit the Firebase Admin SDK Service Agent account.
  * **Keys** -> **ADD KEY** -> **Create new key** -> **Key type**: **JSON** -> **CREATE**.
  * Download and save the created private key to your computer.
* Go to **IAM & Admin** -> **IAM** in [Google Cloud Platform](https://console.cloud.google.com/iam-admin/iam):
  * Edit the Firebase Admin SDK Service Agent account.
  * **ADD ANOTHER ROLE** -> **Role**: **Project** -> **Editor** -> **SAVE**.
* Go to [API Library](https://console.developers.google.com/apis/library) -> search for **Cloud Testing API** and **Cloud Tool Results API** -> enable them.

You can use this one in the README section :)

File: README.md (hunk 5 above)
Comment (thyrlian) [reply README.md:610]: Because you use env variables here, to prevent name conflict, I suggest adding prefix like below:

```
UI_TEST_APK="--app=demo-debug.apk --test=demo-debug-androidTest.apk" && UI_TEST_TYPE="instrumentation" && UI_TEST_DEVICES="--device model=Pixel2,version=27,locale=en,orientation=portrait" && UI_TEST_RESULTS_DIR="build-result" && UI_TEST_PROJECT="<your_project_id>"
gcloud firebase test android run $UI_TEST_APK $UI_TEST_DEVICES --type=$UI_TEST_TYPE --results-dir=$UI_TEST_RESULTS_DIR --project=$UI_TEST_PROJECT
```

**Comment by JaimeToca** [comment]
Awesome ! Thanks a lot. Doing the changes now :+1:

**Comment by JaimeToca** [comment]
hey @thyrlian , I checked out all the comments in the PR and made the changes yesterday. It would be awesome If you could do a last code review. Again, thanks a lot.

P.D: Should I try to add --no-install-recommends for apt-transport-https ca-certificates gnupg in the Dockerfile?. Not sure if you asked for it. I was looking for the comment... :sweat_smile: :

**Comment by thyrlian** [comment]
Thanks Jaime, I'll review it.

Yes, with `--no-install-recommends` would be great.  But let's leave it as this time, and when you have time, try this option to see if you can still run commands with Firebase Test Lab without any problem, then create an other PR please :)

File: README.md [hunk 9]
```diff
@@ -573,6 +573,80 @@ Don't worry about `adbkey` or `adbkey.pub` under `/.android`, not required.
 
   > Unfortunately it is not possible to pass through a USB device (or a serial port) to a container.
 
+## Firebase Test Lab 
+You can also run your UI tests on [Firebase Test Lab](https://firebase.google.com/docs/test-lab) using emulators and physical devices, to do so first you need to
```
Comment (thyrlian) [inline README.md:577]: emulators **or** physical devices

File: android-sdk/firebase-test-lab/.travis.yml [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+services:
+  - docker
+
+install:
+  - git clone https://github.com/thyrlian/AndroidSDK.git ~/AndroidSDK
+
+script:
+  - docker build -t android-sdk-firebase-test-lab ~/AndroidSDK/android-sdk/firebase-test-lab
+  - docker images
+  - cmd=$(cat ~/AndroidSDK/android-sdk/firebase-test-lab/version_inspector.sh) && docker run -it --rm android-sdk-firebase-test-lab bash -c "$cmd"
```
Comment (thyrlian) [inline android-sdk/firebase-test-lab/.travis.yml:14]: We can't have multiple Travis config for one project, but we combine this one to the main one.

**Review by thyrlian — CHANGES_REQUESTED** [review]

**Comment by JaimeToca** [comment]
@thyrlian  changes done. Got the green from travis, let me know if you think something else should be change :smiley: 

Thanks

File: README.md [hunk 11]
```diff
@@ -573,6 +573,79 @@ Don't worry about `adbkey` or `adbkey.pub` under `/.android`, not required.
...
+* Go to **IAM & Admin** -> **Service Accounts** in [Google Cloud Platform](https://console.cloud.google.com/iam-admin/serviceaccounts):
+    * Edit the Firebase Admin SDK Service Agent account.
+    * **Keys** -> **ADD KEY** -> **Create new key** -> **Key type**: **JSON** -> **CREATE**.
+    * Download and save the created private key to your computer.
+* Go to **IAM & Admin** -> **IAM** in [Google Cloud Platform](https://console.cloud.google.com/iam-admin/iam):
+    * Edit the Firebase Admin SDK Service Agent account.
+    * **ADD ANOTHER ROLE** -> **Role**: **Project** -> **Editor** -> **SAVE**.
+* Go to [API Library](https://console.developers.google.com/apis/library) -> search for **Cloud Testing API** and **Cloud Tool Results API** -> enable them.
+
+* Now let's pull and run the Docker image with Google Cloud SDK
```
Comment (thyrlian) [inline README.md:592]: The first part is talking about GCP & Firebase configuration, and the second part is about up and running in container, thus this should not be the same level as the other bullet points.  Please remove the `*` and probably rephrase it to:

`Now we're able to connect to Firebase Test Lab inside the container and deploy UI tests over there.`

File: README.md [hunk 12]
```diff
@@ -573,6 +573,79 @@ Don't worry about `adbkey` or `adbkey.pub` under `/.android`, not required.
...
+
+```bash
+# pull the image
+docker pull thyrlian/android-sdk-firebase-test-lab
+# run the container passing the service account Json (at least the first time) as external volume
+docker run -d -p 2222:22 -v $(pwd)/sdk:/opt/android-sdk -v /path/on/host/auth.json:/auth.json thyrlian/android-sdk-firebase-test-lab
```
Comment (thyrlian) [inline README.md:598]: Need a bit of adjustment in style.  As I've used style like this in previous part: `<your_project>/gradle/wrapper/gradle-wrapper.properties `, it's better that we could stick to consistency, thus could you please change `/path/on/host/auth.json` to `<your_private_key_dir>/auth.json`?

File: README.md [hunk 13]
```diff
@@ -573,6 +573,79 @@ Don't worry about `adbkey` or `adbkey.pub` under `/.android`, not required.
...
+
+# check the running process and get a interactive shell
+docker exec -it `docker ps -aqf "ancestor=thyrlian/android-sdk-firebase-test-lab"` /bin/bash
+
+# authenticate and create any config that you want
+gcloud auth activate-service-account -q --key-file myservicefile.json
```
Comment (thyrlian) [inline README.md:604]: I think some users (including me) would be happy to read more explanation here:
What is the `myservicefile.json`?  Is the command reading from this key file or saving to this key file?

File: README.md [hunk 14]
```diff
@@ -573,6 +573,79 @@ Don't worry about `adbkey` or `adbkey.pub` under `/.android`, not required.
...
+# you should get a matrix output similar to this one but bigger
+┌───────────────────┬────────────────────┬──────────────────────────────────────┬──────────┬─────────────┬─────────────────────────┬───────────────┐
+│      MODEL_ID     │        MAKE        │              MODEL_NAME              │   FORM   │  RESOLUTION │      OS_VERSION_IDS     │      TAGS     │
+├───────────────────┼────────────────────┼──────────────────────────────────────┼──────────┼─────────────┼─────────────────────────┼───────────────┤
+│ Nexus9            │ HTC                │ Nexus 9                              │ VIRTUAL  │ 2048 x 1536 │ 21,22,23,24,25          │               │
+│ NexusLowRes       │ Generic            │ Low-resolution MDPI phone            │ VIRTUAL  │  640 x 360  │ 23,24,25,26,27,28,29,30 │ beta=30       │
+│ OnePlus3T         │ OnePlus            │ OnePlus 3T                           │ PHYSICAL │ 1920 x 1080 │ 26                      │               │
+└───────────────────┴────────────────────┴──────────────────────────────────────┴──────────┴─────────────┴─────────────────────────┴───────────────┘
+
+# compile your apks both app and testing
```
Comment (thyrlian) [inline README.md:618]: both -> for both
testing -> instrumented tests  (because testing could be also local JVM unit tests, it may cause confusion).

File: README.md [hunk 15]
```diff
@@ -573,6 +573,79 @@ Don't worry about `adbkey` or `adbkey.pub` under `/.android`, not required.
...
+
+# if you want to split your test suite into multiple parallel devices add the command --num-uniform-shards=int 
+# example: 20 test to run in 4 parallel devices (5 test per device) --num-uniform-shards=4
+
+# for more info about other configurations such as network-profile, locale, env variables etc ...  
+# go to https://cloud.google.com/sdk/gcloud/reference/beta/firebase/test/android/run
+```
+
+**Important**: To not lose your configurations copy ~/.config/gcloud to the host machine and
+pass it as external volume the next time you run the container.
```
Comment (thyrlian) [inline README.md:644]: pass it as external volume the next time you run the container -> mount it next time

File: image_publisher.sh [hunk 16]
```diff
@@ -51,25 +53,42 @@ docker push $DOCKER_HUB_ACCOUNT/$MAIN_IMAGE_NAME:$TAG
...
+	sed -i "" "s/$MAIN_IMAGE_NAME:latest/$MAIN_IMAGE_NAME:$TAG/" $SUB_IMAGE_FIREBASE_TEST_LAB_DIR/Dockerfile
 else
-	sed -i"" "s/$MAIN_IMAGE_NAME:latest/$MAIN_IMAGE_NAME:$TAG/" $SUB_IMAGE_DIR/Dockerfile
+	sed -i"" "s/$MAIN_IMAGE_NAME:latest/$MAIN_IMAGE_NAME:$TAG/" $SUB_IMAGE_VNC_DIR/Dockerfile
+	sed -i"" "s/$MAIN_IMAGE_NAME:latest/$MAIN_IMAGE_NAME:$TAG/" $SUB_IMAGE_FIREBASE_TEST_LAB_DIR/Dockerfile
 fi
 
-echo "Building the sub image..."
-docker build -t $SUB_IMAGE_NAME $SUB_IMAGE_DIR
+echo "Building the sub vnc image..."
```
Comment (thyrlian) [inline image_publisher.sh:63]: vnc should be uppercase: `VNC`

File: image_publisher.sh [hunk 17]
```diff
@@ -51,25 +53,42 @@ docker push $DOCKER_HUB_ACCOUNT/$MAIN_IMAGE_NAME:$TAG
...
+docker build -t $SUB_IMAGE_VNC_NAME $SUB_IMAGE_VNC_DIR
 
-echo "Revert the change of the base image tag in the sub image file..."
-git checkout -- $SUB_IMAGE_DIR/Dockerfile
+echo "Revert the change of the base image tag in the sub vnc image file..."
```
Comment (thyrlian) [inline image_publisher.sh:66]: vnc should be uppercase: `VNC`

File: image_publisher.sh [hunk 18]
```diff
@@ -51,25 +53,42 @@ docker push $DOCKER_HUB_ACCOUNT/$MAIN_IMAGE_NAME:$TAG
...
+git checkout -- $SUB_IMAGE_VNC_DIR/Dockerfile
 
-image_id=$(docker images $SUB_IMAGE_NAME | awk '{if (NR!=1) {print $3}}')
+image_id=$(docker images $SUB_IMAGE_VNC_NAME | awk '{if (NR!=1) {print $3}}')
 echo "Built sub image ID is: $image_id"
 
-echo "Tagging the sub image with $TAG..."
-docker tag $image_id $DOCKER_HUB_ACCOUNT/$SUB_IMAGE_NAME:$TAG
+echo "Tagging the sub vnc image with $TAG..."
```
Comment (thyrlian) [inline image_publisher.sh:72]: vnc should be uppercase: `VNC`

File: image_publisher.sh [hunk 19]
```diff
@@ -51,25 +53,42 @@ docker push $DOCKER_HUB_ACCOUNT/$MAIN_IMAGE_NAME:$TAG
...
+docker tag $image_id $DOCKER_HUB_ACCOUNT/$SUB_IMAGE_VNC_NAME:$TAG
 
-echo "Pushing the sub image to Docker Hub..."
-docker push $DOCKER_HUB_ACCOUNT/$SUB_IMAGE_NAME:$TAG
+echo "Pushing the sub vnc image to Docker Hub..."
```
Comment (thyrlian) [inline image_publisher.sh:75]: vnc should be uppercase: `VNC`

File: image_publisher.sh [hunk 20]
```diff
@@ -51,25 +53,42 @@ docker push $DOCKER_HUB_ACCOUNT/$MAIN_IMAGE_NAME:$TAG
...
+docker build -t $SUB_IMAGE_VNC_NAME $SUB_IMAGE_VNC_DIR
 
-echo "Revert the change of the base image tag in the sub image file..."
-git checkout -- $SUB_IMAGE_DIR/Dockerfile
+echo "Revert the change of the base image tag in the sub vnc image file..."
+git checkout -- $SUB_IMAGE_VNC_DIR/Dockerfile
 
-image_id=$(docker images $SUB_IMAGE_NAME | awk '{if (NR!=1) {print $3}}')
+image_id=$(docker images $SUB_IMAGE_VNC_NAME | awk '{if (NR!=1) {print $3}}')
 echo "Built sub image ID is: $image_id"
```
Comment (thyrlian) [inline image_publisher.sh:70]: should be `sub VNC image ID`

File: image_publisher.sh [hunk 21]
```diff
@@ -51,25 +53,42 @@ docker push $DOCKER_HUB_ACCOUNT/$MAIN_IMAGE_NAME:$TAG
...
+docker push $DOCKER_HUB_ACCOUNT/$SUB_IMAGE_VNC_NAME:$TAG
+
+echo "Building the sub Firebase Test Lab image..."
+docker build -t $SUB_IMAGE_FIREBASE_TEST_LAB_NAME $SUB_IMAGE_FIREBASE_TEST_LAB_DIR
+
+echo "Revert the change of the base image tag in the sub Firebase Test Lab image file..."
+git checkout -- $SUB_IMAGE_FIREBASE_TEST_LAB_DIR/Dockerfile
+
+image_id=$(docker images $SUB_IMAGE_FIREBASE_TEST_LAB_NAME | awk '{if (NR!=1) {print $3}}')
+echo "Built sub image ID is: $image_id"
```
Comment (thyrlian) [inline image_publisher.sh:85]: should be `sub Firebase Test Lab image ID`

**Review by thyrlian — CHANGES_REQUESTED** [review]
Hi @JaimeToca, I've done my final review, everything is fine, only a few minor changes required.  Thank you so much!  We can get this merged this weekend! :)

File: README.md (hunk 13 above)
Comment (JaimeToca) [reply README.md:604]: I'll add more info, the command is actually reading from this file

File: .travis.yml [hunk 22]
```diff
@@ -7,5 +7,8 @@ services:
 
 script:
   - docker build -t android-sdk ./android-sdk
+  - docker build -t android-sdk-firebase-test-lab ./android-sdk/firebase-test-lab
   - docker images
   - cmd=$(cat ./version_inspector.sh) && docker run -it --rm android-sdk bash -c "$cmd"
+  - cmd=$(cat ./android-sdk/firebase-test-lab/version_inspector.sh) && docker run -it --rm android-sdk-firebase-test-lab bash -c "$cmd"
```
Comment (JaimeToca) [inline .travis.yml:13]: Do we want firebase-test-lab inside android-sdk, right? @thyrlian 
Because if I remove "./android-sdk" this command will fail :sweat_smile:

Comment (thyrlian) [reply .travis.yml:13]: Correct, sorry it was my fault :P

File: .travis.yml [hunk 23]
```diff
@@ -7,5 +7,8 @@ services:
...
+
```
Comment (thyrlian) [inline .travis.yml:14]: Please remove this empty line

File: README.md [hunk 24]
```diff
@@ -573,6 +573,79 @@ Don't worry about `adbkey` or `adbkey.pub` under `/.android`, not required.
...
+
+# if you want to split your test suite into multiple devices in parallel add the command --num-uniform-shards=int 
+# example: 20 test to run in 4 parallel devices (5 test per device) --num-uniform-shards=4
+
+# for more info about other configurations such as network-profile, locale, env variables etc ...  
+# go to https://cloud.google.com/sdk/gcloud/reference/beta/firebase/test/android/run
+```
+
+**Important**: To not lose your configurations copy ~/.config/gcloud to the host machine and
+mount it next time you run the container.
```
Comment (thyrlian) [inline README.md:644]: You have an accidental line break here

**Review by thyrlian — CHANGES_REQUESTED** [review]
Just reviewed, all good, except two minor formatting issues.  We're ready to merge once they're fixed.  Thank you Jaime!

**Comment by JaimeToca** [comment]
Done @thyrlian srry about it. I think it's ready now. 
Waiting for travis :+1: 

I'll take into account all your comments :smile:  for a future PR, so it will demand less time from you
Cheers,

**Review by thyrlian — APPROVED** [review]
Thanks a bunch mate!
